Implement toggleevents for dialog open/close

keithamus opened the pull request:
#### bd7649ea03d308c55de707c6cf8488dc552d43f9
<pre>
Implement toggleevents for dialog open/close
<a href="https://bugs.webkit.org/show_bug.cgi?id=268160">https://bugs.webkit.org/show_bug.cgi?id=268160</a>

Reviewed by Tim Nguyen.

This adds the `beforetoggle` cancelable event to `show()`/`showModal()`
calls. If `beforetoggle` isn&apos;t cancelled, it will show the dialog and
dispatch a `toggle` event.

Likewise, the close method adds a NON cancelable `beforetoggle` and
`toggle` event.

These events can help developers when prepping a dialog to be shown or
hidden, such as doing page calculations or kicking off animations.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/toggle-events.tentative-expected.txt:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
(WebCore::HTMLDialogElement::close):
(WebCore::HTMLDialogElement::queueDialogToggleEventTask):
* Source/WebCore/html/HTMLDialogElement.h:

Canonical link: <a href="https://commits.webkit.org/289695@main">https://commits.webkit.org/289695@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9c17339045bbe620108b17c29cc6210f19ea4af4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/87715 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/7231 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/42099 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/92580 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/38461 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/89766 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/7612 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/15402 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/67743 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/25465 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/90717 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/5807 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/79364 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/48092 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/5594 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/33766 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/37572 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/75992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/34648 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/94466 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/14883 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/10920 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/76571 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/15137 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/75220 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/75806 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18640 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/20170 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/18603 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/7847 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/14899 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/20201 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/14643 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/18087 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/16425 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nt1m commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -177,4 +198,21 @@ void HTMLDialogElement::setIsModal(bool newValue)
     m_isModal = newValue;
 }
 
+void HTMLDialogElement::queueDialogToggleEventTask(DialogState oldState, DialogState newState)
```
This should be shared somehow with the details / popover code to avoid duplication.

keithamus replied:
I had considered this but I also thought it might be better as a follow up PR. Thoughts?

nt1m replied:
I would prefer refactoring details / popover first and then adopting the new shared code here.

keithamus replied:
https://github.com/WebKit/WebKit/pull/23361

nt1m commented on Source/WebCore/html/HTMLDialogElement.cpp:
```diff
@@ -125,6 +144,8 @@ void HTMLDialogElement::close(const String& result)
 
+    queueDialogToggleEventTask(DialogState::Open, DialogState::Closed);
+
     queueTaskToDispatchEvent(TaskSource::UserInteraction, Event::create(eventNames().closeEvent, Event::CanBubble::No, Event::IsCancelable::No));
```
the close event should be emitted before toggle, to get consistent event ordering (otherwise toggle might fire first when no coalescing is happening, but after when it isn't)

nt1m reviewed (commented):
Please make sure tests are passing before landing

keithamus committed:
Implement toggleevents for dialog open/close
https://bugs.webkit.org/show_bug.cgi?id=268160

Reviewed by Tim Nguyen.

This adds the `beforetoggle` cancelable event to `show()`/`showModal()`
calls. If `beforetoggle` isn't cancelled, it will show the dialog and
dispatch a `toggle` event.

Likewise, the close method adds a NON cancelable `beforetoggle` and
`toggle` event.

These events can help developers when prepping a dialog to be shown or
hidden, such as doing page calculations or kicking off animations.

* LayoutTests/imported/w3c/web-platform-tests/html/semantics/interactive-elements/the-dialog-element/toggle-events.tentative-expected.txt:
* Source/WebCore/html/HTMLDialogElement.cpp:
(WebCore::HTMLDialogElement::show):
(WebCore::HTMLDialogElement::showModal):
(WebCore::HTMLDialogElement::close):
(WebCore::HTMLDialogElement::queueDialogToggleEventTask):
* Source/WebCore/html/HTMLDialogElement.h:

Canonical link: https://commits.webkit.org/289695@main
